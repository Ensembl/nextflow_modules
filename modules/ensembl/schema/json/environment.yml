---
name: "schema_json"
channels:
  - conda-forge
  - bioconda
dependencies:
  - python=3.10
  - pip
  - pip:
    - ensembl-genomio=1.6.0
