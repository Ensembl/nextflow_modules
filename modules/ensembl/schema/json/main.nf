// See the NOTICE file distributed with this work for additional information
// regarding copyright ownership.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

process SCHEMA_JSON {
    tag "$meta.id"
    label 'process_low'

    container "ensemblorg/ensembl-genomio:v1.6.1"

    input:
        tuple val(meta), path(json_file)

    output:
        tuple val(meta), path(json_file), emit: verified_json
        path "versions.yml", emit: versions

    when:
        task.ext.when == null || task.ext.when

    script:
        schema_name = json_file.simpleName
        """
        schemas_json_validate --json_file !{json_file} --json_schema !{schema_name}

        echo -e -n "${task.process}:\n\tensembl-genomio: " > versions.yml
        schemas_json_validate --version >> versions.yml
        """

    stub:
        schema_name = json_file.simpleName
        """
        echo "No change, don't create schema error log"

        echo -e -n "${task.process}:\n\tensembl-genomio: " > versions.yml
        schemas_json_validate --version >> versions.yml
        """
}
