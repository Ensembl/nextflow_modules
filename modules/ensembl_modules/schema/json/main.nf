process SCHEMA_JSON {
    tag "$meta.id"
    label 'process_low'

    container "${ (workflow.containerEngine == 'docker') ?
        'lahcen86/dev_test_genomio:litev3' : '' }"

    input:
        tuple val(meta), path(json_file)

    output:
        tuple val(meta), path(json_file), emit: verified_json
        path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
        schema_name = json_file.simpleName
        """
        schemas_json_validate --json_file !{json_file} --json_schema !{schema_name}

        # Get version from genomio please
        VERSION=\$(python -c "import ensembl.io.genomio; print(ensembl.io.genomio.__version__)")
        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            Genomio: \$VERSION
        END_VERSIONS
        """

    stub:
        schema_name = json_file.simpleName
        """
        echo "No change, don't create schema error log"

        # Get version from genomio please
        VERSION=\$(python -c "import ensembl.io.genomio; print(ensembl.io.genomio.__version__)")
        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            Genomio: \$VERSION
        END_VERSIONS
        """
}
