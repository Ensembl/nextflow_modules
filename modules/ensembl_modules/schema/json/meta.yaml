name: "schema_json"
description: Validate the schema for JSON files
keywords:
  - validate
  - JSON
  - schema
tools:
  - "Schema":
      description: "Nextflow modules for Ensembl pipelines"
      homepage: "https://github.com/Ensembl/nextflow_modules"
      # documentation: ""
      # tool_dev_url: ""
      # doi: ""
      licence: "Apache License 2.0"
input:
  # Only when we have meta
  - meta:
      type: map
      description: |
        Groovy Map containing sample information
        e.g. `[ id:'sample1', single_end:false ]`
  
  - json:
      type: file
      description: JSON file to validate
      pattern: "*.{json}"

output:
  # Only when we have meta
  - meta:
      type: map
      description: |
        Groovy Map containing sample information
        e.g. `[ id:'sample1', single_end:false ]`
  - versions:
      type: file
      description: File containing ensembl-genomio versions
      pattern: "versions.yml"
  - json:
      type: file
      description: verified_json file
      pattern: "*.{json}"

authors:
  - "ensembl-dev@ebi.ac.uk"
maintainers:
  - "ensembl-dev@ebi.ac.uk"