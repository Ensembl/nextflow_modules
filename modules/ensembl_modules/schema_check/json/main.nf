process CHECK_JSON_SCHEMA {
    tag "$meta.id"
    label 'process_low'

    input:
        tuple val(meta), path(json_file)

    output:
        tuple val(meta), path(json_file), emit: verified_json
        path "versions.yml"                , emit: versions

    when:
    task.ext.when == null || task.ext.when

    shell:
        def args = task.ext.args ?: ''
        def prefix = task.ext.prefix ?: "${meta.id}"
        schema_name = json_file.simpleName
        def VERSION = 1
        '''
        schemas_json_validate --json_file !{json_file} --json_schema !{schema_name}
        '''
}


    when:
    task.ext.when == null || task.ext.when

    script:
        def args = task.ext.args ?: ''
        def prefix = task.ext.prefix ?: "${meta.id}"
        integrity_file = "integrity.out"
        def VERSION = 1
        """
        manifest_check_integrity \
            --manifest_file ./manifest.json \
            --no_fail \
            $brc_mode \
            > $integrity_file
        
        # Only keep integrity file if there are errors to report
        if [ ! -s $integrity_file ]
            then rm $integrity_file
        fi


        # Get version from genomio please
        # VERSION=\$(manifest_check_integrity --version)
        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            Genomio: $VERSION
        END_VERSIONS
        """

    stub:
        def args = task.ext.args ?: ''
        def prefix = task.ext.prefix ?: "${meta.id}"
        integrity_file = "integrity.out"
        def VERSION = 1
        """
        echo "No change, don't create manifest error log"

        cat <<-END_VERSIONS > versions.yml
        "${task.process}":
            Genomio: $VERSION
        END_VERSIONS
        """
}
