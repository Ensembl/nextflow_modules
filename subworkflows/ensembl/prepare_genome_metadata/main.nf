// See the NOTICE file distributed with this work for additional information
// regarding copyright ownership.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

include { METADATA_GENOMESUMMARY } from '../../../modules/ensembl/metadata/genomesummary'
include { SCHEMA_JSON } from '../../../modules/ensembl/schema/json'

import groovy.json.JsonSlurper
def readJson(json_path) {
    def slurp = new JsonSlurper()
    def json_file = file(json_path)
    def text = json_file.text
    def not_a_lazy_val = slurp.parseText(text)
    return not_a_lazy_val
}

// function to create meta tuple from genome.json in the form:
// tuple("accession": accession, "production_name": production_name, "prefix": prefix)
def metaFromGenomeJson(json_path) {
    def data = readJson(json_path)

    def prod_name = data.assembly.accession
    def publish_dir = data.assembly.accession

    if ( data.species && data.species.production_name ) {
        prod_name = data.species.production_name
        publish_dir = prod_name
    }
    def has_annotation = false
    if (data.annotation) {
        has_annotation = true
    }

    return [
        accession: data.assembly.accession,
        production_name: prod_name,
        publish_dir: publish_dir,
        prefix: "",
        has_annotation: has_annotation,
    ]
}

workflow PREPARE_GENOME_METADATA {

    take:
        input                   // channel: [ val(meta), path(query_json), path(genome_summary_json) ]

    main:
        channel_versions = Channel.empty()

        METADATA_GENOMESUMMARY(input)
            .meta_json
            .map{ meta, json_file -> tuple(meta + metaFromGenomeJson(json_file), json_file) }
            .set { genome_metadata }
        
        SCHEMA_JSON(genome_metadata)
            .verified_json
            .set { verified_genome_json }

        channel_versions.mix(METADATA_GENOMESUMMARY.versions, SCHEMA_JSON.versions)

    emit:
        prepared_metadata = verified_genome_json // channel: [ val(meta), path(json) ]
        versions = channel_versions      // channel: [ path(versions.yml), path(versions.yml) ]
}

