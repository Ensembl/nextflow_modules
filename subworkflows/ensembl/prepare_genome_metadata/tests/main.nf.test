nextflow_workflow {

    name "Test Subworkflow PREPARE_GENOME_METADATA"
    script "../main.nf"
    workflow "PREPARE_GENOME_METADATA"

    tag "subworkflows"
    tag "subworkflows_ensembl"
    tag "subworkflows/prepare_genome_metadata"

    test("genomesummary workflow test (full test)") {

        when {

            workflow {
                """
                input[0] = [
                    [accession: 'GCA_000001215.4'],
                    file("${projectDir}/tests/modules/ensembl/metadata/genomesummary/meta-query.json", checkIfExists: true),
                    file("${projectDir}/tests/modules/ensembl/metadata/genomesummary/ncbi_genome-summary.json", checkIfExists: true)
                    ]
                """
            }
        }

            then {
                assert snapshot(
                    (workflow).out.prepared_metadata.collect {
                        it.collect { it instanceof String && file(it).exists() ? file(it).name : it }
                        }
                    ).match("genome.json")

            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow).match()},
                { assert snapshot(workflow.out.versions) != null }

            )
        }
    }
}
