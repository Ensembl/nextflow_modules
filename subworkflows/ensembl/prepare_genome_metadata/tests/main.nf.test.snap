{
    "genomesummary workflow test (full test)": {
        "content": [
            {
                "stderr": [
                    "\u001b[33mNextflow 24.10.5 is available - Please consider updating your version to it\u001b(B\u001b[m"
                ],
                "errorReport": null,
                "exitStatus": 0,
                "out": {
                    "0": [
                        [
                            {
                                "accession": "GCA_000001215.4",
                                "production_name": "GCA_000001215.4",
                                "publish_dir": "GCA_000001215.4",
                                "prefix": "",
                                "has_annotation": true
                            },
                            "genome.json:md5,9f34e758f4c9504a8f95476989af31c0"
                        ]
                    ],
                    "1": [
                        "versions.yml:md5,889507ae0439c4584d12cf10d79e9202",
                        "versions.yml:md5,9ce2ad4d825e8f06623e4d566fa7c6bc"
                    ],
                    "prepared_metadata": [
                        [
                            {
                                "accession": "GCA_000001215.4",
                                "production_name": "GCA_000001215.4",
                                "publish_dir": "GCA_000001215.4",
                                "prefix": "",
                                "has_annotation": true
                            },
                            "genome.json:md5,9f34e758f4c9504a8f95476989af31c0"
                        ]
                    ],
                    "versions": [
                        "versions.yml:md5,889507ae0439c4584d12cf10d79e9202",
                        "versions.yml:md5,9ce2ad4d825e8f06623e4d566fa7c6bc"
                    ]
                },
                "failed": false,
                "stdout": [
                    
                ],
                "errorMessage": null,
                "trace": {
                    "tasksFailed": 0,
                    "tasksCount": 2,
                    "tasksSucceeded": 2
                },
                "name": "workflow",
                "success": true
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-04-01T14:21:41.078165261"
    },
    "genome.json": {
        "content": [
            [
                [
                    {
                        "accession": "GCA_000001215.4",
                        "production_name": "GCA_000001215.4",
                        "publish_dir": "GCA_000001215.4",
                        "prefix": "",
                        "has_annotation": true
                    },
                    "genome.json"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.3"
        },
        "timestamp": "2025-04-01T15:09:04.471254559"
    }
}